packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -993,6 +1056,270 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
+    final TextSelection newSelection = TextSelection.collapsed(offset: characterBoundary);
+    _setTextEditingValue(
+      TextEditingValue(text: textBefore + textAfter, selection: newSelection),
+      cause,
+    );
+  }
+
+  /// Deletes a word backwards from the current selection.
+  ///
+  /// If the [selection] is collapsed, deletes a word before the cursor.
+  ///
+  /// If the [selection] is not collapsed, deletes the selection.
+  ///
+  /// {@macro flutter.rendering.RenderEditable.cause}
+  ///
+  /// See also:
+  ///
+  ///   * [deleteForwardByWord], which is same but in the opposite direction.
+  void deleteByWord(SelectionChangedCause cause) {
+    assert(_selection != null);
+
+    if (_readOnly || !_selection!.isValid) {
+      return;
+    }
+
+    if (!_selection!.isCollapsed) {
+      return _deleteSelection(_selection!, cause);
+    }
+
+    // When the text is obscured, the whole thing is treated as one big word.
+    if (obscureText) {
+      return _deleteToStart(_selection!, cause);
+    }
+
+    final String text = textSelectionDelegate.textEditingValue.text;
+    String textBefore = _selection!.textBefore(text);
+    if (textBefore.isEmpty) {
+      return;
+    }
+
+    final int characterBoundary = _getLeftByWord(_textPainter, textBefore.length, false);
+    textBefore = textBefore.trimRight().substring(0, characterBoundary);
+
+    final String textAfter = _selection!.textAfter(text);
+    final TextSelection newSelection = TextSelection.collapsed(offset: characterBoundary);
+    _setTextEditingValue(
+      TextEditingValue(text: textBefore + textAfter, selection: newSelection),
+      cause,
+    );
+  }
+
+  /// Deletes a line backwards from the current selection.
+  ///
+  /// If the [selection] is collapsed, deletes a line before the cursor.
+  ///
+  /// If the [selection] is not collapsed, deletes the selection.
+  ///
+  /// {@macro flutter.rendering.RenderEditable.cause}
+  ///
+  /// See also:
+  ///
+  ///   * [deleteForwardByLine], which is same but in the opposite direction.
+  void deleteByLine(SelectionChangedCause cause) {
+    assert(_selection != null);
+
+    if (_readOnly || !_selection!.isValid) {
+      return;
+    }
+
+    if (!_selection!.isCollapsed) {
+      return _deleteSelection(_selection!, cause);
+    }
+
+    // When the text is obscured, the whole thing is treated as one big line.
+    if (obscureText) {
+      return _deleteToStart(_selection!, cause);
+    }
+
+    final String text = textSelectionDelegate.textEditingValue.text;
+    String textBefore = _selection!.textBefore(text);
+    if (textBefore.isEmpty) {
+      return;
+    }
+
+    // When there is a line break, line delete shouldn't do anything
+    final bool isPreviousCharacterBreakLine = textBefore.codeUnitAt(textBefore.length - 1) == 0x0A;
+    if (isPreviousCharacterBreakLine) {
+      return;
+    }
+
+    final TextSelection line = _getLineAtOffset(TextPosition(offset: textBefore.length - 1));
```

Comment:
Sorry not sure I fully understood what you meant here

Reply:
Ah sorry. I meant is it possible to get do this instead:
`final TextSelection line = _getLineAtOffset(selection.baseOffset)`, to avoid the `final bool isPreviousCharacterBreakLine = textBefore.codeUnitAt(textBefore.length - 1) == 0x0A;` check?

Reply:
Ah I think you should check the affinity of `selection.baseOffset` first, and call `_getLineAtOffset(TextPosition(offset: textBefore.length - 1));` or `_getLineAtOffset(TextPosition(offset: textBefore.length + 1));` accordingly

Reply:
If I understand correctly affinity is only relevant for non-collapsed selections, at this point in the code we're already sure the selection is collapsed. If its not collapsed we call a "deleteSelection" and just delete whatever text is within the selection.

The check for line break is just to be sure we don't delete the line break neither trigger a `_setTextEditingValue` without actually changing any text or selection.

Eg:

When I have the cursor at the beginning of a line:
```
My
|Multiline <-- collapsed selection at the start of this line
Text
```
If I press cmd+delete it shouldnt change anything.

Reply:
I think affinity matters to caret positioning too. In your example that's only true if the affinity is `TextAffinity.downstream`. When the affinity is upstream the caret should appear at the end of the first line.

Reply:
Oh the documentation says "This affects only line breaks caused by wrapping, not explicit newline characters"

Reply:
Oh ok! I'll look in more detail into it!

Reply:
But anyway, when you get the `TextPosition` of a newline character I would expect the engine to always return `downstream` so checking affinity should still work.

Reply:
While investigating the affinity stuff I compared the way line selection and line delete are working and I found some inconsistencies between them and inconsistencies with TextEdit (which is one of the sources we've been using to check the expected behaviour).

* Place cursor at the start of a line without line breaks:
  * **Selection**: It selects the previous line
  * **Delete**: it deletes the previous line
  * **TextEdit**: it doesn't delete or select the previous line
* Place cursor at the start of a line with line breaks:
  * **Selection**: It selects the previous line
  * **Delete**: it doesn't delete anything
  * **TextEdit**: it doesn't delete or select the previous line

Here is a confusing recording where I try to show what I mean:
https://user-images.githubusercontent.com/1267623/112716882-e625fd80-8ee0-11eb-8083-0ee768a7e414.mov

What should we do in the context of this PR. Make delete consistent with Selection or with TextEdit?

cc @justinmc

Reply:
Good catch!  I've opened an issue for the selection problem.  It should match TextEdit.  https://github.com/flutter/flutter/issues/79291

I think you should ideally make delete do the same behavior as TextEdit.

However, if that requires you to change some internal methods in RenderEditable that would also affect selection, then we can put it off until another PR and solve both delete and selection at the same time later, after this PR.

So it's up to you, but the goal is definitely to match TextEdit in all cases.

Reply:
Yeah I think it would be better to have a separated PR dedicated to tackle this particular issue in both selections and delete actions. To make sure we reuse as most code as possible and make the PR easy to review. 

Is it ok?

Reply:
Yes, I'm on board with that!